lib/specdiff/hashprint.rb

```diff
@@ -0,0 +1,154 @@
+# Tries to print nested hash/array structures such that they are as convenient
+# as possible for a diffing algorithm designed for text (like source code).
+# Basically tries to reproduce consistent literal source code form for ruby
+# hashes and arrays and objects (although it falls back to #inspect).
+class Specdiff::Hashprint
+  def self.call(...)
+    new.call(...)
+  end
+
+  INDENTATION_SPACES = 2
+  SPACE = " ".freeze
+  COMMA = ",".freeze
+  NEWLINE = "\n".freeze
+
+  def call(thing)
+    @indentation_level = 0
+    @indentation_per_level = SPACE * INDENTATION_SPACES
+    @indent = ""
+    @skip_next_opening_indent = false
+
+    @output = StringIO.new
+
+    output(thing)
+
+    @output.string
+  end
+
+private
+
+  def recalculate_indent
+    @indent = @indentation_per_level * @indentation_level
+  end
+
+  def increase_indentation
+    @indentation_level += 1
+    recalculate_indent
+  end
+
+  def decrease_indentation
+    @indentation_level -= 1
+    recalculate_indent
+  end
+
+  def with_indentation_level(temporary_level)
+    old_level = @indentation_level
+    @indentation_level = temporary_level
+    recalculate_indent
+
+    yield
+
+    @indentation_level = old_level
+    recalculate_indent
+  end
+
+  def skip_next_opening_indent
+    @skip_next_opening_indent = true
+
+    nil
+  end
+
+  def this_indent_should_be_skipped
+    if @skip_next_opening_indent
+      @skip_next_opening_indent = false
+      true
+    else
+      false
+    end
+  end
+
+  # #=== allows us to rely on Module implementing #=== instead of relying on the
+  # thing (which could be any kind of wacky object) having to implement
+  # #is_a? or #kind_of?
+  def output(thing)
+    if Hash === thing
+      output_hash(thing)
+    elsif Array === thing
+      output_array(thing)
+    else
+      output_unknown(thing)
+    end
+  end
+
+  HASH_OPEN = "{".freeze
+  HASH_CLOSE = "}".freeze
+  HASHROCKET = "=>".freeze
+  COLON = ":".freeze
+
+  def output_hash(hash)
+    @output << @indent unless this_indent_should_be_skipped
+
+    @output << HASH_OPEN
+    # unless hash.empty?
+      @output << NEWLINE
+
+      increase_indentation
+      hash.each do |key, value|
```

Comment:
maybe sort keys, for this kind of example:

![image](https://github.com/odinhb/specdiff/assets/17880055/042214b5-c97c-4b9f-bf27-dc8fba192e09)

Reply:
Maybe add a config